sycl/include/sycl/ext/oneapi/experimental/work_group_memory.hpp

```diff
@@ -0,0 +1,103 @@
+//===-------------------- work_group_memory.hpp ---------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <type_traits>
+
+namespace sycl {
+inline namespace _V1 {
+namespace detail {
+template <typename T> struct is_unbounded_array : std::false_type {};
+
+template <typename T> struct is_unbounded_array<T[]> : std::true_type {};
+
+template <typename T>
+inline constexpr bool is_unbounded_array_v = is_unbounded_array<T>::value;
+
+class work_group_memory_impl {
+public:
+  work_group_memory_impl() : wgm_size{0}, buffer_size{0} {}
+  work_group_memory_impl(const work_group_memory_impl &rhs) = default;
+  work_group_memory_impl &
+  operator=(const work_group_memory_impl &rhs) = default;
+  work_group_memory_impl(size_t wgm_size, size_t buffer_size)
+      : wgm_size{wgm_size}, buffer_size{buffer_size} {}
+  size_t wgm_size;
+  size_t buffer_size;
+};
+
+inline size_t getWorkGroupMemoryOwnSize(detail::work_group_memory_impl *wgm) {
+  return wgm->wgm_size;
+}
+
+// The following 3 functions help us get the address of the first element of a
+// multi-dimensional array, be it bounded or unbounded. A scalar is also
+// included. In that case, it just returns the address of the scalar.
+template <typename DataT> auto getData(DataT &scalar) { return &scalar; }
+
+template <typename DataT, size_t N> auto getData(DataT (&bounded_arr)[N]) {
+  return getData(bounded_arr[0]);
+}
+
+template <typename DataT> auto getData(DataT (&unbounded_arr)[]) {
+  return getData(unbounded_arr[0]);
+}
+
+} // namespace detail
+
+namespace ext::oneapi::experimental {
+template <typename DataT, typename PropertyListT = empty_properties_t>
+class __SYCL_SPECIAL_CLASS __SYCL_TYPE(work_group_memory) work_group_memory
+    : sycl::detail::work_group_memory_impl {
+public:
+  using value_type = std::remove_all_extents_t<DataT>;
+
+private:
+  using decoratedPtr = typename sycl::detail::DecoratedType<
+      DataT, access::address_space::local_space>::type *;
```

Comment:
This looks suspicious.  Rather than creating a decorated pointer to `DataT`, I think this should create a decorated pointer to `value_type`.  In the case when `DataT` is an array type, I think you want `decoratedPtr` to be a pointer to the element type, not a pointer to the array type.

This is possibly the cause of the problem with the SPIRV translator which happens when `DataT` is an unbounded array.

Reply:
If you do that, can it return `ptr` as a `DataT &` without a `reinterpret_cast`? I worry that would require UB through strict aliasing violation.

Reply:
Yes, I agree it would require a `reinterpret_cast` in the implementation of `operator DataT&()`.

It would be interesting to know if this avoids the problem with the SPIRV translator, though.  We we still have the problem anyway because of the `reinterpret_cast` back to an unbounded array type?

Reply:
`reinterpret_cast` of array of T to/from T is not UB. An aggregate containing T is considered compatible with T and may alias T.

Reply:
Does that also apply to `T* -> T[]`? If so, I suppose it would be fine.